test/baddtlstest.c

```diff
@@ -0,0 +1,603 @@
+
+    if (data + 3 > dataend)
+        return 0;
+    /* Check CCS message content, including seq# 2 */
+    if (data[0] != 0x01 || data[1] != 0x00 || data[2] != 0x02)
+        return 0;
+    data += 3; /* DTLS1_BAD_VER CCS length */
+
+    /* We expect the Finished message here.*/
+    if (data + DTLS1_RT_HEADER_LENGTH > dataend)
+        return 0;
+    /* Check record header type, version */
+    if (data[0] != 0x16 || data[1] != 0x01 || data[2] != 0x00)
+        return 0;
+    /* Check epoch is now 1 */
+    if (data[3] != 00 || data[4] != 0x01)
+        return 0;
+
+    return 1;
+}
+
+static int send_data_packet(BIO *rbio, int seq)
+{
+    static unsigned char data_hdr[] = {
+        0x17, /* Application data */
+        0x01, 0x00, /* DTLS1_BAD_VER */
+        0x00, 0x01, /* Epoch */
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, /* Seq# */
+        0x00, 0x30, /* Length */
+    };
+    static unsigned char record_iv[16];
+    static unsigned char data_msg[0x20] = "Packet x";
+
+    data_hdr[10] = seq;
+    sprintf((char *)data_msg, "Packet %d", seq);
+
+    /* We want to use ssl3_cbc_digest_record() here but it's not
+     * available if we're testing against a shared library incarnation
+     * of OpenSSL. So we do it ourselves. */
+    hmac_packet(data_hdr, data_msg, 9);
+
+    /* Pad */
+    memset(data_msg + 29, 0x02, 3);
+    encrypt_packet(record_iv, data_msg, sizeof(data_msg));
+
+    BIO_write(rbio, data_hdr, sizeof(data_hdr));
+    BIO_write(rbio, record_iv, sizeof(record_iv));
+    BIO_write(rbio, data_msg, sizeof(data_msg));
+
+    return 1;
+}
+
+int main(int argc, char *argv[])
+{
+    SSL_SESSION *sess;
+    SSL_CTX *ctx;
+    SSL *con;
+    BIO *rbio;
+    BIO *wbio;
+    BIO *err;
+    int testresult = 0;
+    char buf[10];
+    int ret;
+
+    err = BIO_new_fp(stderr, BIO_NOCLOSE | BIO_FP_TEXT);
+
+    CRYPTO_set_mem_debug(1);
+    CRYPTO_mem_ctrl(CRYPTO_MEM_CHECK_ON);
+
+    RAND_bytes(session_id, sizeof(session_id));
+    RAND_bytes(master_secret, sizeof(master_secret));
+    RAND_bytes(cookie, sizeof(cookie));
+    RAND_bytes(server_random + 4, sizeof(server_random) - 4);
+    time((void *)server_random);
+
+    sess = client_session();
+    if (!sess) {
+        printf("Failed to generate SSL_SESSION\n");
+        goto end;
+    }
+
+    handshake_md = EVP_MD_CTX_new();
+    if (!handshake_md || !EVP_DigestInit_ex(handshake_md, EVP_md5_sha1(), NULL)) {
+        printf("Failed to initialise handshake_md\n");
+        goto end;
+    }
+
+    ctx = SSL_CTX_new(DTLS_client_method());
+    if (!ctx) {
+        printf("Failed to allocate SSL_CTX\n");
+        goto end_md;
+    }
+    if (!SSL_CTX_set_min_proto_version(ctx, DTLS1_BAD_VER)) {
+        printf("SSL_CTX_set_min_proto_version() failed\n");
+        goto end_ctx;
+    }
+    if (!SSL_CTX_set_max_proto_version(ctx, DTLS1_BAD_VER)) {
+        printf("SSL_CTX_set_max_proto_version() failed\n");
+        goto end_ctx;
+    }
+
+    SSL_CTX_set_read_ahead(ctx, 1);
```

Comment:
Do you need this? `read_ahead` should not need to be set in DTLS...it should be used anyway.